File: Content/Code/queue-circular-array/cpp/circular_queue_exercise.cpp
#include <iostream>
using namespace std;

class IntQueue {
private:
    int* A;
    int capacity;
    int frontIndex;
    int count;

public:
    IntQueue(int cap) {
        capacity = cap;
        A = new int[capacity];
        frontIndex = 0;
        count = 0;
    }

    ~IntQueue() {
        delete[] A;
    }

    bool isEmpty() {
        // TODO
        return false;
    }

    bool isFull() {
        // TODO
        return false;
    }

    bool enqueue(int value) {
        // TODO
        return false;
    }

    int dequeue() {
        // TODO
        return -1;
    }

    int front() {
        // TODO
        return -1;
    }

    int size() {
        // TODO
        return -1;
    }
};

void checkAtLine(int actual, int expected, int line, const char* expression) {
    if (actual == expected) {
        cout << "PASS at test line " << line << " (" << expression << "): got " << actual << endl;
    } else {
        cout << "FAIL at test line " << line << " (" << expression
             << "): expected " << expected << " but got " << actual << endl;
    }
}

void checkAtLine(bool actual, bool expected, int line, const char* expression) {
    if (actual == expected) {
        cout << "PASS at test line " << line << " (" << expression << "): got " << actual << endl;
    } else {
        cout << "FAIL at test line " << line << " (" << expression
             << "): expected " << (expected ? "true" : "false")
             << " but got " << (actual ? "true" : "false") << endl;
    }
}

#define check(actual, expected) checkAtLine((actual), (expected), __LINE__, #actual)

void testQueue() {
    IntQueue queue(4);

    check(queue.isEmpty(), true);
    check(queue.isFull(), false);
    check(queue.dequeue(), -1);
    check(queue.front(), -1);

    check(queue.enqueue(4), true);
    check(queue.enqueue(7), true);
    check(queue.enqueue(9), true);

    check(queue.dequeue(), 4);
    check(queue.dequeue(), 7);

    check(queue.enqueue(2), true);
    check(queue.enqueue(5), true);
    check(queue.enqueue(8), true);
    check(queue.isFull(), true);
    check(queue.enqueue(10), false);

    check(queue.front(), 9);
    check(queue.dequeue(), 9);
    check(queue.dequeue(), 2);
    check(queue.dequeue(), 5);
    check(queue.dequeue(), 8);

    check(queue.isEmpty(), true);
    check(queue.size(), 0);
    check(queue.dequeue(), -1);

    check(queue.enqueue(11), true);
    check(queue.front(), 11);

    IntQueue wrapped(257);
    bool wrappedOk = true;
    for (int round = 0; round < 20; round++) {
        for (int i = 0; i < 257; i++) wrappedOk = wrapped.enqueue(round * 257 + i) && wrappedOk;
        for (int i = 0; i < 257; i++) wrappedOk = (wrapped.dequeue() == round * 257 + i) && wrappedOk;
    }
    check(wrappedOk && wrapped.isEmpty(), true);
}

int main() {
    testQueue();
    return 0;
}

File: Content/Code/queue-circular-array/python/circular_queue_exercise.py
import inspect


class IntQueue:
    def __init__(self, capacity):
        self.A = [None] * capacity
        self.front_index = 0
        self.count = 0

    def is_empty(self):
        # TODO
        return False

    def is_full(self):
        # TODO
        return False

    def enqueue(self, value):
        # TODO
        return False

    def dequeue(self):
        # TODO
        return -1

    def front(self):
        # TODO
        return -1

    def size(self):
        # TODO
        return -1


def check(actual, expected):
    line = inspect.currentframe().f_back.f_lineno
    if actual == expected:
        print(f"PASS at test line {line}: got {actual!r}")
    else:
        print(f"FAIL at test line {line}: expected {expected!r} but got {actual!r}")


def test_queue():
    queue = IntQueue(4)

    check(queue.is_empty(), True)
    check(queue.is_full(), False)
    check(queue.dequeue(), -1)
    check(queue.front(), -1)

    check(queue.enqueue(4), True)
    check(queue.enqueue(7), True)
    check(queue.enqueue(9), True)

    check(queue.dequeue(), 4)
    check(queue.dequeue(), 7)

    check(queue.enqueue(2), True)
    check(queue.enqueue(5), True)
    check(queue.enqueue(8), True)
    check(queue.is_full(), True)
    check(queue.enqueue(10), False)

    check(queue.front(), 9)
    check(queue.dequeue(), 9)
    check(queue.dequeue(), 2)
    check(queue.dequeue(), 5)
    check(queue.dequeue(), 8)

    check(queue.is_empty(), True)
    check(queue.size(), 0)
    check(queue.dequeue(), -1)

    check(queue.enqueue(11), True)
    check(queue.front(), 11)

    wrapped = IntQueue(257)
    wrapped_ok = True
    for round_number in range(20):
        for i in range(257):
            wrapped_ok = wrapped.enqueue(round_number * 257 + i) and wrapped_ok
        for i in range(257):
            wrapped_ok = wrapped.dequeue() == round_number * 257 + i and wrapped_ok
    check(wrapped_ok and wrapped.is_empty(), True)


test_queue()
